File: packages/native/runtime-kotlin/src/main/kotlin/com/pyreon/runtime/PyreonJson.kt
// PyreonJson — serialization helper for the `<WebView>` live-data bridge
// (mirror of PyreonJSON.swift). PMTC emits `PyreonJson.encode(signal)`
// for `<WebView data={signal}>`; this encodes any `@Serializable` value
// (PMTC-emitted data classes carry `@Serializable`) to a compact JSON
// string for injecting into the hosted page as `window.__pyreonData`.
//
// Uses the same `Json.encodeToString(serializer<T>(), value)` surface as
// PyreonStorage, so it links against the same kotlinx-serialization API
// (covered by the runtime's kotlinc verify stubs).

package com.pyreon.runtime

import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer

object PyreonJson {
    inline fun <reified T> encode(value: T): String =
        Json.encodeToString(serializer<T>(), value)
}

File: packages/native/runtime-swift/Sources/PyreonRuntime/PyreonJSON.swift
// PyreonJSON — the serialization helper for the `<WebView>` live-data
// bridge. PMTC emits `PyreonJSON.encode(signal)` for `<WebView
// data={signal}>`; this encodes any Encodable value (PMTC-emitted structs
// are `Codable`) to a compact JSON string suitable for injecting into the
// hosted page as `window.__pyreonData`. Never throws into the view layer —
// returns the JSON literal `null` on the (practically unreachable) encode
// failure, which the page reads as "no data yet".

import Foundation

public enum PyreonJSON {
    public static func encode<T: Encodable>(_ value: T) -> String {
        guard let data = try? JSONEncoder().encode(value),
              let json = String(data: data, encoding: .utf8)
        else { return "null" }
        return json
    }
}
